src/frontends/pytorch/src/op/scatter.cpp

```diff
@@ -17,58 +17,114 @@ namespace op {
     // aten::scatter.value_reduce(Tensor self, int dim, Tensor index, Scalar value, *, str reduce) -> Tensor:
+    // aten::scatter.value_out(Tensor self, int dim, Tensor index, Scalar value, *, Tensor(a!) out) -> Tensor(a!)
+    // aten::scatter.value_reduce_out(Tensor self, int dim, Tensor index, Scalar value, *, str reduce, Tensor(a!) out)
+    // -> Tensor(a!)
 
     // Inplace schema
     // aten::scatter_.value(Tensor(a!) self, int dim, Tensor index, Scalar value) -> Tensor(a!):
     // aten::scatter_.src(Tensor(a!) self, int dim, Tensor index, Tensor src) -> Tensor(a!):
     // aten::scatter_.reduce(Tensor(a!) self, int dim, Tensor index, Tensor src, *, str reduce) -> Tensor(a!):
     // aten::scatter_.value_reduce(Tensor(a!) self, int dim, Tensor index, Scalar value, *, str reduce) -> Tensor(a!):
-    num_inputs_check(context, 4, 5);
+    num_inputs_check(context, 4, 6);
     auto input = context.get_input(0);
     auto dim = context.get_input(1);
     auto index = context.mark_node(std::make_shared<v0::Convert>(context.get_input(2), element::i32));
     auto src = context.get_input(3);
 
     auto reduction = v12::ScatterElementsUpdate::Reduction::NONE;
     auto input_num = context.get_input_size();
-    if (input_num > 4 && !context.input_is_none(input_num - 1)) {
-        auto reduce_mode = context.const_input<std::string>(input_num - 1);
-        if (reduce_mode == "add") {
-            reduction = v12::ScatterElementsUpdate::Reduction::SUM;
-        } else if (reduce_mode == "multiply") {
-            reduction = v12::ScatterElementsUpdate::Reduction::PROD;
-        }
+    // 5 argument can be reduction represened as string or out represented as Tensor
+    if (input_num > 4 && !context.input_is_none(4) && context.get_input_type(4).is<type::Str>()) {
+        auto reduce_mode = context.const_input<std::string>(4);
+        reduction = get_reduction_mode(reduce_mode);
     }
-    auto src_partial_shape = src.get_partial_shape();
-    auto index_shape_rank = get_shape_rank(context, index);
-    auto index_shape = std::get<0>(index_shape_rank);
-    auto index_rank = std::get<1>(index_shape_rank);
 
-    // Source input can be either Tensor which should be passed in original shape or Scalar that should be broadcasted
-    // into shape of indices.
-    // TODO: Figure out way to dynamically broadcast scalar src only, without affecting Tensor src. Current
-    // implementation will fail if Scalar source would have dynamic rank.
-    if (src_partial_shape.rank().is_static() && src_partial_shape.rank().get_length() == 0) {
-        src = context.mark_node(std::make_shared<v3::Broadcast>(src, index_shape));
+    auto src_input_dtype = prepare_source(context, src, index, input);
+    auto res =
+        context.mark_node(std::make_shared<v12::ScatterElementsUpdate>(input, index, src_input_dtype, dim, reduction));
+    if (input_num == 6 || (input_num == 5 && reduction == v12::ScatterElementsUpdate::Reduction::NONE)) {
```

Comment:
What if reduction was provided and resulted as none? I guess spec doesn't allow that, but safer would be to check again here that input is not `type::Str`.

Reply:
it is not possible to provide none as string on torch side,
reduce ([str](https://docs.python.org/3/library/stdtypes.html#str), optional) – reduction operation to apply, can be either 'add' or 'multiply'.

attempt set it to none lead to error

Reply:
I understand but pytorch can change behaivior or it can make it none by default, it will be safe to check for string again here

Reply:
ok, used check for string here